### PR TITLE
Reimplement stream_tweets()

### PR DESCRIPTION
**Pull request by hadley**
This is now much closer to the "metal" than previously, using curl directly in order to access its connection interface. It now reads in chunks of 64kb, only working with complete lines, and only recording actual tweets to the output file.

Other UI improvements:

* A progress bar reports number of tweets downloaded and download rate

* An interrupt (Esc/Ctrl+C) will now terminate the stream, while still returning data

* Defaults to writing on top of existing file so that you can continue to stream more tweets into one file.

### CODE REVIEW

**Comment by hadley** [comment]
@llrs ok this is really truly the last big thing I'll do on rwteet this week

**Comment by jthomasmock** [comment]
Just played around with this branch, and most everything seems to be working as intended. 

Notes:
- I did have to manually load a previously created "proper" Twitter OAuth token, whereas `search_tweet()`, `post_tweet()`, etc just "work".
- The OAuth handshake tried to kick off the first time I ran `stream_tweets()` with the default `token = NULL`, but after confirming the Twitter bot in browser, and being redirected back to R/RStudio, `stream_tweets()` then failed to work interactively (it did write a token to temp, but that token gives a 401 unauthorized error)
  - I don't do much with OAuth, so this could be user error on my part, however the other functions in the package "just work" (ie `search_tweets()`, `post_tweet()`, etc
- The streaming works great after using a token, I ran several tests of various hashtags, wrote to the default `stream_tweets.json` as well as named files.
- I'd love to see a running total of file size as opposed to or in addition to the download rate. I was surprised at getting 5 MB files for only ~700 tweets.

File: R/stream.R [hunk 1]
```diff
@@ -1,443 +1,230 @@
...
+  stop_time <- Sys.time() + timeout
+  
+  n_seen <- 0
   if (verbose) {
-    message(
-      paste0("Streaming tweets for ", timeout, " seconds...")
+    pb <- progress::progress_bar$new(
+      total = NA,
+      show_after = 0,
+      format = "Streaming tweets: :n tweets written / :rate / :elapsedfull"
```
Comment (llrs) [inline R/stream.R:95]: Adding a \n at the end of the message will help with the display that @jthomasmock comments. But will create a new message every 10 seconds or so.

File: R/stream.R [hunk 2]
```diff
@@ -1,443 +1,230 @@
...
-    restapi = FALSE,
-    query,
-    param = params
-  )
-  ## temp file if none provided
   if (is.null(file_name)) {
-    tmp <- TRUE
-    file_name <- tmp_json()
+    file_name <- "stream_tweets.json"
+    inform(paste0("Writing to '", file_name, "'"))
```
Comment (llrs) [inline R/stream.R:63]: It could be confusing if people set append = TRUE and see later "Writing to". Perhaps check that the file exists and then adjust the message to "Appending to ...."?
Also the message should be given independently whether the user provides a file name or not.

**Review by llrs — APPROVED** [review]
Not sure if you have finished on this or not. But looks good to me. I only left two comments but this could get merged

File: R/stream.R (hunk 1 above)
Comment (hadley) [reply R/stream.R:95]: I can just add a separate counter for number of bytes.

File: R/stream.R (hunk 2 above)
Comment (hadley) [reply R/stream.R:63]: Agreed with appending vs writing, but I think this pattern of only messaging when the user doesn't supply an argument is useful. See https://design.tidyverse.org/def-inform.html for details.

**Comment by abuchmueller** [comment]
Love the new implementation of `stream_tweets()`. I've been playing around with it since I'm developing a package which depends on `rtweet` but I'm finding the deprecation of `parse_stream()` troublesome. I've run into multiple issues with running `jsonlite::stream_in()` when opening a file

1. Can't open in viewer

```
df <- jsonlite::stream_in(file("file.json"))
df <- tibble::as_tibble(df)
View(df)
```

returns `Error in View : 'names' attribute [1] must be the same length as the vector [0]`. 
I've tested this with multiple json files from different rtweet versions but can't find the issue. 

2. Resulting data frame differs vastly from when using `parse_stream()` in shape and expected behaivior

when using `jsonlite::stream_in()`
```
# A tibble: 2,073 x 36
   created_at       id id_str  text       display_text_ra… source    truncated in_reply_to_sta… in_reply_to_sta…
   <chr>         <dbl> <chr>   <chr>      <list>           <chr>     <lgl>                <dbl> <chr>           
 1 Tue Mar 09… 1.37e18 136930… "@newsmax… <int [2]>        "<a href… FALSE              1.37e18 136914718813633…
 2 Tue Mar 09… 1.37e18 136930… "RT @Edwi… <NULL>           "<a href… FALSE             NA       NA              
 3 Tue Mar 09… 1.37e18 136930… "Joes #No… <NULL>           "<a href… FALSE             NA       NA              
 4 Tue Mar 09… 1.37e18 136930… "RT @jimt… <NULL>           "<a href… FALSE             NA       NA              
 5 Tue Mar 09… 1.37e18 136930… "RT @CGTN… <NULL>           "<a href… FALSE             NA       NA              
 6 Tue Mar 09… 1.37e18 136930… "RT @Judi… <NULL>           "<a href… FALSE             NA       NA              
 7 Tue Mar 09… 1.37e18 136930… "@GreggJa… <int [2]>        "<a href… TRUE               1.37e18 136928259050347…
 8 Tue Mar 09… 1.37e18 136930… "RT @Rich… <NULL>           "<a href… FALSE             NA       NA              
 9 Tue Mar 09… 1.37e18 136930… "RT @Emer… <NULL>           "<a href… FALSE             NA       NA              
10 Tue Mar 09… 1.37e18 136930… "RT @Tony… <NULL>           "<a href… FALSE             NA       NA              
# … with 2,063 more rows, and 27 more variables: in_reply_to_user_id <dbl>, in_reply_to_user_id_str <chr>,
#   in_reply_to_screen_name <chr>, user <df[,39]>, geo <lgl>, coordinates <lgl>, place <df[,9]>,
#   contributors <lgl>, is_quote_status <lgl>, quote_count <int>, reply_count <int>, retweet_count <int>,
#   favorite_count <int>, entities <df[,5]>, favorited <lgl>, retweeted <lgl>, filter_level <chr>, lang <chr>,
#   timestamp_ms <chr>, retweeted_status <df[,34]>, quoted_status_id <dbl>, quoted_status_id_str <chr>,
#   quoted_status <df[,32]>, quoted_status_permalink <df[,3]>, possibly_sensitive <lgl>,
#   extended_tweet <df[,4]>, extended_entities <df[,1]>
```

when using `parse_stream()`

```
# A tibble: 2,073 x 90
   user_id   status_id  created_at          screen_name text           source  display_text_wi… reply_to_status…
   <chr>     <chr>      <dttm>              <chr>       <chr>          <chr>              <dbl> <chr>           
 1 13395068… 136930361… 2021-03-09 15:06:36 PakapronSo… "@newsmax A p… Twitte…              128 136914718813633…
 2 172912017 136930361… 2021-03-09 15:06:36 ddbak69     "Illegal immi… Twitte…               NA NA              
 3 172912017 136930434… 2021-03-09 15:09:31 ddbak69     "As recently … Twitte…               NA NA              
 4 12976149… 136930361… 2021-03-09 15:06:36 Andreas439… "Joes #NoJoke… Twitte…               NA NA              
 5 27622250… 136930361… 2021-03-09 15:06:36 PrayagGordy "President Tr… Twitte…               NA NA              
 6 10370033… 136930361… 2021-03-09 15:06:36 CEEC86_Shu… "12 states su… Twitte…               NA NA              
 7 19181618… 136930361… 2021-03-09 15:06:36 jjwills2    "There’s a cr… Twitte…               NA NA              
 8 75630932… 136930361… 2021-03-09 15:06:36 jpula23687  "@GreggJarret… Twitte…              126 136928259050347…
 9 13509532… 136930361… 2021-03-09 15:06:36 TMichaelBy… "Suggested qu… Twitte…               NA NA              
10 87905794… 136930361… 2021-03-09 15:06:36 Pamwindham1 "In less than… Twitte…               NA NA              
# … with 2,063 more rows, and 82 more variables: reply_to_user_id <chr>, reply_to_screen_name <chr>,
#   is_quote <lgl>, is_retweet <lgl>, favorite_count <int>, retweet_count <int>, quote_count <int>,
#   reply_count <int>, hashtags <list>, symbols <list>, urls_url <list>, urls_t.co <list>,
#   urls_expanded_url <list>, media_url <list>, media_t.co <list>, media_expanded_url <list>,
#   media_type <list>, ext_media_url <list>, ext_media_t.co <list>, ext_media_expanded_url <list>,
#   ext_media_type <chr>, mentions_user_id <list>, mentions_screen_name <list>, lang <chr>,
#   quoted_status_id <chr>, quoted_text <chr>, quoted_created_at <dttm>, quoted_source <chr>,
#   quoted_favorite_count <int>, quoted_retweet_count <int>, quoted_user_id <chr>, quoted_screen_name <chr>,
#   quoted_name <chr>, quoted_followers_count <int>, quoted_friends_count <int>, quoted_statuses_count <int>,
#   quoted_location <chr>, quoted_description <chr>, quoted_verified <lgl>, retweet_status_id <chr>,
#   retweet_text <chr>, retweet_created_at <dttm>, retweet_source <chr>, retweet_favorite_count <int>,
#   retweet_retweet_count <int>, retweet_user_id <chr>, retweet_screen_name <chr>, retweet_name <chr>,
#   retweet_followers_count <int>, retweet_friends_count <int>, retweet_statuses_count <int>,
#   retweet_location <chr>, retweet_description <chr>, retweet_verified <lgl>, place_url <chr>,
#   place_name <chr>, place_full_name <chr>, place_type <chr>, country <chr>, country_code <chr>,
#   geo_coords <list>, coords_coords <list>, bbox_coords <list>, status_url <chr>, name <chr>, location <chr>,
#   description <chr>, url <chr>, protected <lgl>, followers_count <int>, friends_count <int>,
#   listed_count <int>, statuses_count <int>, favourites_count <int>, account_created_at <dttm>,
#   verified <lgl>, profile_url <chr>, profile_expanded_url <chr>, account_lang <lgl>,
#   profile_banner_url <chr>, profile_background_url <chr>, profile_image_url <chr>
```

which in itself wouldn't be a big issue (except for me having to rewrite some of my package) but the contents seem to differ between dataframes e.g.

dataframe from `parse_stream()`:
`sum(df$is_retweet == TRUE)`
returns `[1] 1509`

vs.

data frame from `jsonlite::stream_in()`:
`sum(df$retweeted)`
returns `[1] 0`

or

df from `parse_stream()`:
```
which(df$is_retweet == FALSE)
  [1]    1    4    8   15   18   20   21   22   23   29   33   35   42   43   45   49   56   57   58   59   60
 [22]   61   62   63   64   65   66   68   69   78   82   89   92   95   98  106  107  111  118  128  133  134
 [43]  148  149  151  180  181  189  191  193  210  211  213  217  232  240  246  264  267  271  273  276  277
 [64]  278  279  281  282  286  287  289  296  302  303  310  312  315  317  318  322  323  328  329  330  331
 [85]  336  337  340  341  343  347  348  349  350  351  356  357  358  362  363  365  366  376  378  380  382
[106]  384  387  389  395  396  399  400  427  428  429  432  433  434  435  436  448  450  456  460  461  462
[127]  465  466  467  468  473  474  476  477  478  479  488  492  518  522  523  535  539  542  544  545  546
[148]  550  551  554  557  558  564  566  568  571  577  579  588  589  590  596  597  599  600  617  622  625
[169]  633  635  646  648  652  656  665  667  668  677  681  682  689  690  692  693  696  701  704  706  707
[190]  709  717  718  719  722  723  730  732  738  743  745  746  750  760  761  769  770  771  772  777  778
[211]  779  781  785  788  789  790  791  792  796  801  803  806  807  809  813  818  829  833  836  841  842
[232]  846  851  852  854  861  862  866  874  877  897  905  906  907  910  918  929  931  935  936  939  943
[253]  945  947  950  953  955  956  960  964  967  973  975  981  983  984  985  987  989 1008 1012 1014 1016
[274] 1023 1034 1049 1050 1051 1053 1055 1056 1059 1060 1063 1070 1079 1080 1083 1086 1087 1093 1095 1103 1111
[295] 1112 1126 1131 1139 1140 1151 1158 1159 1174 1176 1177 1178 1179 1180 1181 1183 1188 1190 1191 1193 1194
[316] 1195 1202 1210 1211 1218 1225 1230 1239 1241 1242 1248 1255 1257 1260 1275 1276 1279 1281 1282 1284 1286
[337] 1288 1290 1297 1298 1310 1311 1313 1328 1331 1335 1336 1337 1341 1347 1348 1350 1351 1352 1354 1357 1361
[358] 1362 1364 1370 1376 1385 1388 1397 1400 1402 1404 1405 1410 1411 1413 1414 1416 1421 1426 1431 1442 1445
[379] 1449 1454 1458 1463 1465 1469 1470 1479 1484 1489 1491 1493 1495 1497 1498 1499 1506 1508 1511 1514 1516
[400] 1519 1523 1526 1530 1537 1539 1541 1542 1545 1547 1550 1552 1553 1554 1556 1557 1564 1581 1589 1591 1595
[421] 1596 1600 1604 1611 1615 1617 1621 1623 1624 1626 1627 1629 1635 1639 1640 1641 1642 1645 1648 1652 1653
[442] 1658 1667 1669 1670 1672 1678 1681 1687 1690 1691 1692 1693 1694 1697 1703 1704 1720 1721 1722 1738 1739
[463] 1742 1743 1744 1750 1756 1759 1762 1763 1764 1765 1766 1767 1771 1774 1775 1776 1779 1780 1781 1783 1784
[484] 1791 1792 1793 1810 1818 1819 1822 1827 1830 1832 1834 1836 1838 1846 1848 1854 1861 1864 1866 1868 1869
[505] 1870 1875 1876 1878 1881 1886 1888 1889 1892 1897 1898 1903 1904 1905 1907 1909 1916 1919 1920 1921 1924
[526] 1933 1937 1939 1941 1945 1949 1950 1955 1958 1963 1966 1972 1976 1979 1980 1982 1984 1991 1996 2004 2005
[547] 2016 2017 2025 2026 2027 2031 2034 2035 2039 2046 2051 2054 2057 2059 2060 2061 2065 2067
```

vs. df from `jsonlite::stream_in()`

```
which(zx$retweeted_status$retweeted == FALSE)
   [1]    2    4    5    6    8    9   10   11   12   13   15   16   18   22   23   24   26   27   29   31   32
  [22]   33   34   35   38   40   41   43   46   49   50   51   52   53   55   56   58   59   61   62   64   66
  [43]   67   69   70   71   72   73   74   77   78   79   80   82   83   84   85   87   88   89   90   91   92
  [64]   93   94   95   97   98   99  102  103  104  105  106  107  108  110  112  113  114  115  116  117  118
  [85]  119  120  121  122  123  124  125  126  127  128  131  132  133  134  135  136  138  140  142  143  144
 [106]  145  146  147  148  149  150  151  152  153  154  155  156  159  161  162  163  165  166  167  168  169
 [127]  170  171  172  173  174  175  176  178  179  180  181  182  183  185  186  187  188  190  191  192  193
 [148]  194  195  196  197  198  199  200  201  202  203  204  205  206  207  209  210  212  214  216  219  222
 [169]  223  226  228  229  230  231  232  233  235  236  238  239  240  243  244  245  246  248  250  251  253
 [190]  256  257  258  261  262  263  264  265  268  270  271  272  273  275  276  279  281  282  283  287  288
 [211]  289  290  291  295  296  298  301  302  303  304  305  306  307  309  311  313  315  317  318  320  322
 [232]  323  324  327  330  331  332  333  334  335  336  337  338  339  340  341  342  343  344  347  353  354
 [253]  355  356  357  358  359  360  361  362  363  364  366  368  369  370  372  373  376  377  378  381  384
 [274]  385  386  387  390  395  396  397  398  399  400  402  403  405  406  407  408  409  410  411  412  413
 [295]  414  415  416  417  418  419  420  421  423  424  425  426  427  428  431  432  433  434  435  436  437
 [316]  438  439  441  442  443  445  446  448  453  454  455  457  458  459  461  462  463  466  467  468  469
 [337]  471  472  474  476  477  479  480  481  482  484  486  487  488  489  490  491  492  493  497  498  499
 [358]  500  502  505  506  507  508  509  510  511  512  513  514  515  517  518  520  521  523  524  525  526
 [379]  528  529  531  532  533  534  535  536  537  538  539  541  543  544  545  547  549  550  551  552  553
 [400]  554  555  556  557  559  560  563  564  565  566  567  568  570  571  574  575  576  577  580  583  584
 [421]  586  587  588  589  590  592  593  595  598  599  601  602  603  604  605  606  607  610  611  614  615
 [442]  616  617  618  619  620  622  624  625  627  628  629  632  633  634  636  637  638  639  640  641  642
 [463]  643  646  647  648  649  650  651  653  654  659  660  661  662  666  668  669  670  672  673  674  677
 [484]  681  682  684  685  686  687  688  690  692  693  696  698  699  700  701  702  704  705  706  707  709
 [505]  710  711  712  713  714  715  716  718  719  720  722  723  724  726  727  728  729  733  734  736  737
 [526]  739  740  741  742  745  746  748  749  750  751  752  753  755  756  758  759  760  762  763  764  765
 [547]  766  767  768  769  770  771  772  773  774  775  777  778  779  780  781  785  786  788  789  790  791
 [568]  792  793  794  796  797  798  799  800  801  802  804  806  807  810  811  812  813  814  817  818  820
 [589]  822  823  825  826  827  829  830  832  835  836  837  838  839  841  842  843  845  846  848  849  850
 [610]  851  853  855  856  857  858  859  861  865  866  868  870  871  872  873  874  875  876  877  878  879
 [631]  880  881  882  883  885  886  887  888  890  892  893  894  895  896  897  898  900  901  902  903  904
 [652]  905  906  907  908  909  911  912  913  914  915  916  918  919  920  921  922  923  924  925  926  927
 [673]  931  933  936  937  938  940  942  943  945  946  947  948  949  951  952  953  954  955  956  958  959
 [694]  962  963  965  966  969  970  971  972  973  975  977  978  979  980  981  983  984  985  986  987  990
 [715]  991  992  993  994  995  996  997  998  999 1001 1002 1003 1004 1006 1007 1008 1009 1010 1013 1014 1015
 [736] 1016 1018 1019 1020 1021 1023 1025 1026 1027 1028 1031 1032 1033 1034 1035 1036 1038 1039 1040 1041 1042
 [757] 1043 1050 1052 1053 1054 1055 1057 1058 1059 1061 1064 1065 1069 1070 1071 1072 1073 1074 1075 1076 1079
 [778] 1080 1081 1082 1083 1084 1086 1087 1089 1090 1091 1092 1093 1094 1095 1097 1098 1099 1100 1102 1103 1104
 [799] 1105 1106 1107 1109 1110 1111 1112 1113 1115 1116 1119 1120 1121 1122 1123 1124 1126 1127 1128 1129 1130
 [820] 1132 1134 1135 1136 1139 1140 1141 1142 1143 1144 1145 1146 1147 1148 1149 1150 1151 1152 1153 1156 1157
 [841] 1158 1160 1163 1164 1166 1168 1170 1171 1172 1173 1175 1176 1177 1178 1179 1180 1181 1184 1185 1186 1187
 [862] 1188 1189 1190 1191 1192 1193 1194 1195 1196 1197 1198 1200 1202 1204 1205 1206 1207 1208 1209 1210 1211
 [883] 1212 1213 1214 1215 1216 1217 1219 1220 1222 1223 1224 1226 1230 1231 1232 1234 1235 1236 1237 1238 1239
 [904] 1242 1243 1246 1248 1249 1251 1252 1253 1254 1255 1257 1259 1260 1261 1262 1264 1265 1266 1267 1268 1270
 [925] 1271 1272 1273 1274 1275 1276 1277 1278 1280 1281 1283 1284 1285 1286 1287 1288 1289 1290 1291 1293 1294
 [946] 1295 1297 1298 1300 1303 1304 1305 1306 1309 1312 1313 1315 1316 1317 1318 1320 1321 1322 1323 1325 1326
 [967] 1328 1329 1331 1332 1333 1334 1335 1336 1337 1338 1340 1341 1343 1345 1347 1348 1349 1350 1351 1353 1355
 [988] 1356 1357 1359 1361 1362 1364 1365 1366 1367 1368 1370 1371 1372
 [ reached getOption("max.print") -- omitted 509 entries ]
```


So in short, when using `jsonlite::stream_in()` it seems there are not retweets in the data while when using `parse_stream() ` two thirds of the data are retweets.

I've uploaded the [json file in question](https://github.com/ropensci/rtweet/files/6125460/tweetsaboutbiden.json.zip), with which you can reproduce and check for yourselves. I would be glad to hear from someone, hoping it's just an issue with my code and not `rtweet`.

**Comment by hadley** [comment]
@abuchmueller instead of commenting on a closed PR, can you please file a new issue? (preferably with a reprex)

**Comment by abuchmueller** [comment]
@hadley alright, here it is #534
